F-FrontEnd/src/run_tests.py

```diff
@@ -0,0 +1,782 @@
+                    line = line.lower()
+                    # Find module declarations and uses
+                    module_decl_line = reg_module_decl.match(line)
+                    submodule_decl_line = reg_submodule_decl.match(line)
+                    use_module_line = reg_use_module.match(line)
+                    if module_decl_line is not None:
+                        mod_name = module_decl_line.group(1)
+                        if mod_name not in ('procedure', 'function', 'subroutine'):
+                            # assert mod_name not in mod_to_file, 'Module "%s" already defined in "%s"' % \
+                            #                                    (mod_name, mod_to_file[mod_name])
+                            mod_to_file[mod_name] = test_case
+                            this_testcase_mods.add(mod_name)
+                            if debug_output: print('\t' + mod_name)
+                    elif submodule_decl_line is not None:
+                        submod_short_name = submodule_decl_line.group(2)
+                        parent_mod_name = reg_space.sub('', submodule_decl_line.group(1))
+                        submod_name = '%s:%s' % (parent_mod_name, submod_short_name)
+                        mod_to_file[submod_name] = test_case
+                        this_testcase_mods.add(submod_name)
+                        if parent_mod_name not in this_testcase_mods:
+                            parent_mod_file = mod_to_file[parent_mod_name]
+                            deps.add(parent_mod_file)
+                    elif use_module_line is not None:
+                        mod_name = use_module_line.group(1)
+                        # assert mod_name in mod_to_file, 'Module "%s" not defined in this or previous modules' % mod_name
+                        if mod_name not in this_testcase_mods:
+                            mod_file = mod_to_file.get(mod_name)
+                            # assert mod_file is not None, 'Module not declared'
+                            if mod_file is not None:
+                                deps.add(mod_file)
+                if deps:
+                    if debug_output: print('\tdependencies:')
+                    for dep in sorted(deps):
+                        if debug_output: print('\t\t' + dep)
+        testcase_dep_lst = {}
+        # Convert dependency trees into lists
+        for test_case, direct_deps in testcase_deps.items():
+            lst = []
+            test_case_direct_deps = testcase_deps[test_case]
+            visited = set()
+            for dir_dep in test_case_direct_deps:
+                if dir_dep in visited:
+                    continue
+                for dep in testcase_dep_lst[dir_dep]:
+                    if dep not in visited:
+                        visited.add(dep)
+                        lst.append(dep)
+                if dir_dep not in visited:
+                    visited.add(dir_dep)
+                    lst.append(dir_dep)
```

Comment:
```suggestion
                assert dir_dep not in visited:
                visited.add(dir_dep)
                lst.append(dir_dep)
```
If the assertion fails, the dependency graph is cyclic.

Reply:
I've tested it: both options happen (in visited and not) an seem legit. Please clarify what you mean.

Reply:
Yea, I misunderstood what these dependencies mean. So cyclic dependencies are legitimate.
Then the `if` statement is correct.